### PR TITLE
Hailem/116 implement df image component

### LINKED ISSUE #116: [results] Implement df-image component
The component will have the following attributes: 
- with (auto, <number>%, <number>)
- height (auto, <number>%, <number>)
- fit (scale-down, ...)

### PR DESCRIPTION
**Pull request by hailem**
fixes #116

### CODE REVIEW

File: packages/results/src/image.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  private _element_styles: GenericObject = [];
+
+  public static get styles(): CSSResultArray {
+    return [...super.styles, ...styles];
+  }
+
+  static get observedAttributes(): string[] {
+    return [...super.observedAttributes, 'height-ratio', 'background-color'];
+  }
```
Comment (carlosescri) [inline packages/results/src/image.ts]: Remove this, all properties are converted from attributes so they're already observed. If you set `reflect: true` in the property options then they will be converted to attributes when the property value changes.

File: packages/results/src/image.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @example
+ *
+ * ```html
+ * <df-image src="/images/logo.jpg"></df-icon>
+ * ```
+ *
+ * @public
+ */
+@customElement('df-image')
```
Comment (carlosescri) [inline packages/results/src/image.ts]: Don't use the decorator (see #121).

File: packages/results/src/image.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  public attributeChangedCallback(name: string, old: string | null, value: string | null): void {
+    if (name == 'height-ratio') {
+      this.style.setProperty('padding-bottom', value);
+    }
+    if (name == 'background-color') {
+      this.style.setProperty('background-color', value);
+    }
+    this._element_styles[name] = value;
+    super.attributeChangedCallback(name, old, value);
+  }
```
Comment (carlosescri) [inline packages/results/src/image.ts]: I think you don't need this, you can compose the styles object in the `render` method. LitElement knows when something changed and it has to re-render so you shouldn't need it.

File: packages/results/src/image.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * ```html
+ * <df-image src="/images/logo.jpg"></df-icon>
+ * ```
+ *
+ * @public
+ */
+@customElement('df-image')
+export class Image extends Widget {
+  @property({ type: String }) public name = '';
```
Comment (carlosescri) [inline packages/results/src/image.ts]: Unused?

File: packages/results/src/image.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  @property({ type: String }) public src = '';
+  @property({ type: String }) public width = '';
+  @property({ type: String }) public height_ratio = '';
+  @property({ type: String }) public aspect_ratio = '1:1';
```
Comment (carlosescri) [inline packages/results/src/image.ts]: Unused?

File: packages/results/src/image.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ * ```
+ *
+ * @public
+ */
+@customElement('df-image')
+export class Image extends Widget {
+  @property({ type: String }) public name = '';
+  @property({ type: String }) public src = '';
+  @property({ type: String }) public width = '';
+  @property({ type: String }) public height_ratio = '';
```
Comment (carlosescri) [inline packages/results/src/image.ts]: Use camel case for property names. You can set the attribute name in the property options:

```ts
@property({ type: String, attribute: 'height-ratio' }) public heightRatio;
```

Also, don't set empty strings as default values. If a required attribute is missing you can throw an exception or just render nothing (don't return anything in the render function).

File: packages/results/src/image.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  @property({ type: String }) public aspect_ratio = '1:1';
+
+  private _element_styles: GenericObject = [];
+
```
Comment (carlosescri) [inline packages/results/src/image.ts]: Add this for #121:

```ts
public static get tag(): string {
  return 'df-image';
}
```

File: packages/results/src/image.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    super.attributeChangedCallback(name, old, value);
+  }
+
+  /** @public */
+  public render() {
+    return html`
+      <img src="${this.src}" style="${styleMap(this._element_styles)}" />
+    `;
+  }
+}
```
Comment (carlosescri) [inline packages/results/src/image.ts]: I would do something like this:

```ts
private _buildStyles(): GenericObject {
  // build styles here
}

public render() {
  return html`
    <img src="${this.src}" style="${styleMap(this._buildStyles())}" />
  `;
}
```

File: packages/results/src/image.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+ * <df-image src="/images/logo.jpg"></df-icon>
+ * ```
+ *
+ * @public
+ */
+@customElement('df-image')
+export class Image extends Widget {
+  @property({ type: String }) public name = '';
+  @property({ type: String }) public src = '';
+  @property({ type: String }) public width = '';
```
Comment (carlosescri) [inline packages/results/src/image.ts]: Unused?

Comment (carlosescri) [reply packages/results/src/image.ts]: The `height` property is missing. If any of them are set, they should fix width and height.

File: packages/results/src/image.ts (hunk 6 above)
Comment (carlosescri) [reply packages/results/src/image.ts]: The `fit` property (or `image-fit`) is missing, it should allow setting `object-fit` in the CSS.

```ts
export type ImageFit = 'contain' | 'fill' | 'cover' | 'none' | 'scale-down';
// …
@property({ type: String, attribute: 'image-fit' }) public fit: ImageFit = 'contain';
```

Take into account that you can set CSS based on attributes, see my comments in the styles file.

File: packages/results/src/styles/image.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    :host img {
+      width: calc(100%);
+      height: calc(100%);
+      position: absolute;
+      top: 0;
+      left: 0;
+      object-fit: contain;
+    }
```
Comment (carlosescri) [inline packages/results/src/styles/image.ts]: This CSS is for the case there's a _height ratio_ defined so you could do:

```css
:host([height-ratio]) {
  // rules for that case…
}

:host([height-ratio]) img {
  // …
}
```

For the `fit` (or `image-fit`) attribute, you could do:

```css
:host([image-fit="contain"]) img { object-fit: contain; }
:host([image-fit="fill"]) img { object-fit: fill; }
/* and so on */
```

So CSS is managed by HTML attributes and not by JS, when possible.

File: packages/results/src/card.ts [hunk 11]
```diff
@@ -27,9 +27,7 @@ const DEFAULT_TEMPLATE = function(html: Function, _ctx: GenericObject) {
         ></df-price>
       </div>
     </div>
-    <div class="media">
-      <img src="//via.placeholder.com/300x225?text=Product%20Image" ?hidden=${!this.data.image_link} />
-    </div>
+    <df-image src="${this.data.image_link}"></df-image>
```
Comment (carlosescri) [inline packages/results/src/card.ts]: Don't remove the `?hidden` property, all widgets have `hidden` and `visuallyhidden` available (if you take into account, as you did, `super.styles` in the styles getter).

For the moment we don't have a _placeholder_ for images (we could add it if you think it's useful) so we just hide them if there's no image…

File: packages/results/src/image.ts (hunk 1 above)
Comment (hailem) [reply packages/results/src/image.ts]: done

File: packages/results/src/image.ts (hunk 2 above)
Comment (hailem) [reply packages/results/src/image.ts]: done

File: packages/results/src/image.ts (hunk 7 above)
Comment (hailem) [reply packages/results/src/image.ts]: done

File: packages/results/src/image.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
+/**
+ * Material Design Icons as custom elements.
+ *
+ * @packageDocumentation
+ */
```
Comment (carlosescri) [inline packages/results/src/image.ts]: Remove copy & paste!

**Comment by carlosescri** [comment]
**Default use case**
Gray background is not part of the image and was customized with a CSS variable.

image

**Aspect ratio of 150%**

By default image fits as `scale-down` (the _best of_ `contain` and `none`):

image

Other options (in order) are `contain`, `cover`, and `fill`:

image

**Fixed Height**

image

The same with `cover`, `fill` and `contain` as values for the `fit` attribute:

image

**Pull request merged**